idea/idea-frontend-fir/src/org/jetbrains/kotlin/idea/fir/evaluate/FirCompileTimeConstantEvaluator.kt

```diff
@@ -0,0 +1,193 @@
+            ConstantValueKind.String -> CompileTimeType.STRING
+
+            else -> CompileTimeType.BYTE
+        }
+    }
+
+    // Unary operators
+    private fun FirConstExpression<*>.evaluate(function: FirSimpleFunction): FirConstExpression<*>? {
+        if (value == null) return null
+        return evalUnaryOp(
+            function.name.asString(),
+            kind.toCompileTimeType(),
+            value!!
+        )?.let {
+            it.toConstantValueKind()?.toConstExpression(source, it)
+        }
+    }
+
+    // Binary operators
+    private fun FirConstExpression<*>.evaluate(
+        function: FirSimpleFunction,
+        other: FirConstExpression<*>
+    ): FirConstExpression<*>? {
+        if (value == null || other.value == null) return null
+        return evalBinaryOp(
+            function.name.asString(),
+            kind.toCompileTimeType(),
+            value!!,
+            other.kind.toCompileTimeType(),
+            other.value!!
+        )?.let {
+            it.toConstantValueKind()?.toConstExpression(source, it)
+        }
+    }
+
+    ////// KINDS
+
+    private fun FirTypeRef.toConstantValueKind(): ConstantValueKind<out Number>? =
+        when (this) {
+            !is FirResolvedTypeRef -> null
+            !is FirImplicitBuiltinTypeRef -> type.toConstantValueKind()
+            is FirImplicitIntTypeRef -> ConstantValueKind.Int
+            else -> null
+        }
+
+    private fun ConeKotlinType.toConstantValueKind(): ConstantValueKind<out Number>? =
+        when (this) {
+            is ConeKotlinErrorType -> null
+            is ConeLookupTagBasedType -> {
+                when (lookupTag.name.asString()) {
+                    "Byte" -> ConstantValueKind.Byte
+                    "Double" -> ConstantValueKind.Double
+                    "Float" -> ConstantValueKind.Float
+                    "Int" -> ConstantValueKind.Int
+                    "Long" -> ConstantValueKind.Long
+                    "Short" -> ConstantValueKind.Short
+                    else -> null
+                }
+            }
+            is ConeFlexibleType -> upperBound.toConstantValueKind()
+            is ConeCapturedType -> lowerType?.toConstantValueKind() ?: constructor.supertypes!!.first().toConstantValueKind()
+            is ConeDefinitelyNotNullType -> original.toConstantValueKind()
+            is ConeIntersectionType -> intersectedTypes.first().toConstantValueKind()
+            is ConeStubType -> null
+            is ConeIntegerLiteralType -> null
+        }
+
+    // TODO: better to remove `out Number` restriction
+    private fun <T : Any> T.toConstantValueKind(): ConstantValueKind<out Number>? =
+        when (this) {
+            is Byte -> ConstantValueKind.Byte
+            is Double -> ConstantValueKind.Double
+            is Float -> ConstantValueKind.Float
+            is Int -> ConstantValueKind.Int
+            is Long -> ConstantValueKind.Long
+            is Short -> ConstantValueKind.Short
```

Comment:
Is Char, String and Boolean missing here?

Reply:
Ditto. Initially, every conversion in this evaluator was limited to `Number` (I don't remember why I did that, maybe just as a starting point). Now those were converted as well.